From 7db3760431ae2967d4852bdcbe6114636a74350b Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Wed, 24 May 2023 15:55:24 +0800
Subject: [PATCH 190/263] [Huawei] userswap: mark swap-out buffer PTE as
 writable

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I6CAIM

--------------------------------

Add a VM_WRITE check for the swap-out buffer. If the swap-out buffer VMA
contains VM_WRITE, the PTE should be marked as writable.

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 mm/userswap.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/mm/userswap.c b/mm/userswap.c
index ff40a0d82576..ccf22382d264 100644
--- a/mm/userswap.c
+++ b/mm/userswap.c
@@ -215,7 +215,7 @@ static unsigned long vm_insert_anon_page(struct vm_area_struct *vma,
 {
 	struct mm_struct *mm = vma->vm_mm;
 	int ret = 0;
-	pte_t *pte;
+	pte_t *pte, dst_pte;
 	spinlock_t *ptl;
 
 	if (unlikely(anon_vma_prepare(vma)))
@@ -233,7 +233,10 @@ static unsigned long vm_insert_anon_page(struct vm_area_struct *vma,
 	inc_mm_counter(mm, MM_ANONPAGES);
 	reliable_page_counter(page, mm, 1);
 	page_add_new_anon_rmap(page, vma, addr, false);
-	set_pte_at(mm, addr, pte, mk_pte(page, vma->vm_page_prot));
+	dst_pte = mk_pte(page, vma->vm_page_prot);
+	if (vma->vm_flags & VM_WRITE)
+		dst_pte = pte_mkwrite(pte_mkdirty(dst_pte));
+	set_pte_at(mm, addr, pte, dst_pte);
 
 out_unlock:
 	pte_unmap_unlock(pte, ptl);
-- 
2.17.1

